### PR TITLE
Add Uppy Companion to wopi example

### PR DESCRIPTION
**Pull request by dschmidt**
## Description
Add Uppy Companion to wopi example, so cloud imports work as soon as the importer app is shipped with oC Web.

## Motivation and Context
Cloud imports need the Uppy Companion microservice.

## How Has This Been Tested?
Tested locally with my own build of oC Web.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: deployments/examples/ocis_wopi/.env [hunk 1]
```diff
@@ -55,3 +55,8 @@ TIKA_IMAGE=
 # you need uncomment following line. Please see documentation at
 # https://owncloud.dev/ocis/deployment/monitoring-tracing/
 #COMPOSE_FILE=docker-compose.yml:monitoring_tracing/docker-compose-additions.yml
+
+### Uppy Companion
+COMPANION_DOMAIN=
+COMPANION_ONEDRIVE_KEY=
+COMPANION_ONEDRIVE_SECRET=
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply deployments/examples/ocis_wopi/.env:62]: I asked @micbar  and he wanted me to add it to the wopi example, so it's "full featured".

File: deployments/examples/ocis_wopi/docker-compose.yml [hunk 2]
```diff
@@ -3,14 +3,15 @@ version: "3.7"
 
 services:
   traefik:
-    image: traefik:v2.9.1
+    image: traefik:v2.10.1
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply deployments/examples/ocis_wopi/docker-compose.yml:6]: Oh well, fair enough

File: deployments/examples/ocis_wopi/.env (hunk 1 above)
[the remark this answers is not in the document]
Comment (dschmidt) [reply deployments/examples/ocis_wopi/.env:62]: Well, OneDrive integration obviously won't ... but e.g. https://uppy.io/docs/url/ would

[the remark this answers is not in the document]
Comment (dschmidt) [reply deployments/examples/ocis_wopi/.env:62]: ... we most likely will need to add a setting to enable certain integrations in the ui, but the way the import is configured is about to change anyway, so it's probably not worth adding it now.

[the remark this answers is not in the document]
Comment (dschmidt) [reply deployments/examples/ocis_wopi/.env:62]: Hm yeah, well - I thought that was kinda self explanatory, no other variables in `.env` have that kind of comment although they are empty and everything works

**Comment by tbsbdr** [comment]
@micbar I'd like to reference the wopi deployment example in the releasenotes, but would not drive documentation further for this feature. ok for you?

**Comment by micbar** [comment]
I am still not decided if we should deploy it in the continouus deployments at all.

It was a big effort doing it for the example deployment.

**Comment by dschmidt** [comment]
We could also just deploy the public link provider...

**Comment by mmattel** [comment]
@micbar SonarCloud seems to be "unwilling" to continue, not only here, because I also tried with a rebase and it "hung". I think the important two are green...

**Comment by micbar** [comment]
Example is incomplete. Not sufficient. We need to check what we can do with the config file templating.

**Comment by nicholas-wilson-au** [comment]
This is highlighted in our release notes https://doc.owncloud.com/docs/next/ocis_release_notes.html and is still open.  We should remove this from the release notes if there is no way for our customers to use this feature.

**Comment by marciofeld** [comment]
I tried to deploy the Upload Companion in my test environment, but nothing changed on the Web Interface. I'm pretty sure I made everything correct.

Is that working with v4.0.0+?

**Comment by dschmidt** [comment]
> I tried to deploy the Upload Companion in my test environment, but nothing changed on the Web Interface. I'm pretty sure I made everything correct.
> 
> Is that working with v4.0.0+?

I'm sorry, this PR was probably misleading. We changed the way the importer is configured after this PR was sent and then noone made a decision how to progress here, so it got stale and outdated.
In the long run the env var might do the right thing again, until then you need to provide your own `config.json` - and have a `companionUrl` setting there.

```json
{
...
  "external_apps": [
    {
      "id": "importer",
      "path": "web-app-importer",
      "config": {
        "companionUrl": "https://your.companion.url/",
        "webdavCloudType": "owncloud"
      }
    }
  ]
...
}
```

**Comment by mmattel** [comment]
@dschmidt can you fix that PR so it can be merged and the example becomes available?

**Comment by dschmidt** [comment]
I can, but I need a decision we actually want that. 

Quoting @micbar :
> I am still not decided if we should deploy it in the continouus deployments at all.

**Comment by mmattel** [comment]
This is somehow rediculous.
On the one hand we add it to the changelog
Then we provide a PR making it avaiable, but the PR is somehow not complete
A user asks about a working example
We provide that here
And then we discuss about making the PR complete for deployment
I dont get that ...?

**Comment by micbar** [comment]
@mmattel Like i said there is a technical blocker.

**Comment by marciofeld** [comment]
> > I tried to deploy the Upload Companion in my test environment, but nothing changed on the Web Interface. I'm pretty sure I made everything correct.
> > Is that working with v4.0.0+?
> 
> I'm sorry, this PR was probably misleading. We changed the way the importer is configured after this PR was sent and then noone made a decision how to progress here, so it got stale and outdated. In the long run the env var might do the right thing again, until then you need to provide your own `config.json` - and have a `companionUrl` setting there.
> 
> ```json
> {
> ...
>   "external_apps": [
>     {
>       "id": "importer",
>       "path": "web-app-importer",
>       "config": {
>         "companionUrl": "https://your.companion.url/",
>         "webdavCloudType": "owncloud"
>       }
>     }
>   ]
> ...
> }
> ```

Thanks @dschmidt !
Could you let me know the path of this `config.json` file?

**Comment by dschmidt** [comment]
In the Web dev setup we do it like this:

https://github.com/owncloud/web/blob/b7f1c5dd5ddd7faa62e1fddbb8ae18e67093d011/docker-compose.yml#L16
https://github.com/owncloud/web/blob/b7f1c5dd5ddd7faa62e1fddbb8ae18e67093d011/docker-compose.yml#L69


so basically you need to mount it to your container and use the `WEB_UI_CONFIG_FILE` env var to point oCIS to it
I would recommend to download your config.json from https://your.ocis.url/config.json, then add the section I pasted above and then mount it to your container...

I'm sorry this is so hard right now...

**Comment by marciofeld** [comment]
Thanks @dschmidt ! I successfully managed to deploy the Cloud Import Feature.

I have an OwnCloud v10 remote instance and I tried to import files using a generated Public URL from it, but it didn't work.
I tried with and without Password, but in both cases it returned a error message `Connection with Companion failed`, as you can see in the screenshot below:

image

Please find the latest Companion container log lines below:

```
::ffff:172.31.11.165 - - [19/Oct/2023:21:33:41 +0000] "OPTIONS /webdavPublicLink/list/?publicLinkURL=<PUBLIC_URL_HIDDEN> HTTP/1.1" 204 0 "<INFINITESCALE_URL_HIDDEN>" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36"
companion: 2023-10-19T21:33:42.110Z [info] companion.client.version uppy client version @uppy/companion-client=3.2.0
companion: 2023-10-19T21:33:42.111Z [warn] 8e508542-6949-4229-bf2d-e291d49d5e16 provider.middleware.invalid invalid provider options detected. Provider will not be loaded
::ffff:172.31.11.165 - - [19/Oct/2023:21:33:42 +0000] "GET /webdavPublicLink/list/?publicLinkURL=<PUBLIC_URL_HIDDEN> HTTP/1.1" 400 11 "<INFINITESCALE_URL_HIDDEN>" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36"
```

**Comment by dschmidt** [comment]
> Thanks @dschmidt ! I successfully managed to deploy the Cloud Import Feature.

Cool, good to hear!

> Please find the latest Companion container log lines below:
> 
> ```
> ::ffff:172.31.11.165 - - [19/Oct/2023:21:33:41 +0000] "OPTIONS /webdavPublicLink/list/?publicLinkURL=<PUBLIC_URL_HIDDEN> HTTP/1.1" 204 0 "<INFINITESCALE_URL_HIDDEN>" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36"
> companion: 2023-10-19T21:33:42.110Z [info] companion.client.version uppy client version @uppy/companion-client=3.2.0
> companion: 2023-10-19T21:33:42.111Z [warn] 8e508542-6949-4229-bf2d-e291d49d5e16 provider.middleware.invalid invalid provider options detected. Provider will not be loaded
> ::ffff:172.31.11.165 - - [19/Oct/2023:21:33:42 +0000] "GET /webdavPublicLink/list/?publicLinkURL=<PUBLIC_URL_HIDDEN> HTTP/1.1" 400 11 "<INFINITESCALE_URL_HIDDEN>" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36"
> ```


The error message comes from here: https://github.com/owncloud/uppy/blob/f3ce884e43d6dd8732a0cafb8d32ee8ba9a12d9b/packages/%40uppy/companion/src/server/provider/index.js#L118 -> https://github.com/owncloud/uppy/blob/f3ce884e43d6dd8732a0cafb8d32ee8ba9a12d9b/packages/%40uppy/companion/src/server/provider/index.js#L25

So apparently you don't seem to have `server.host` and `server.protocol` configured. See Companion documentation here: https://uppy.io/docs/companion/#server

**Comment by marciofeld** [comment]
Thanks again @dschmidt !
I've been reading all the Uppy Companion docs and also all the OCIS docker-compose examples.

I'm still in doubt about how to solve this last error since I would like to import data from OwnCloud v10 to Infinite Scale v4, where my OwnCloud relies on EBS Volumes to store data while my Infinite Scale relies on S3 Buckets.

On Companion docs, there are environment variables available for these services: `BOX`, `DROPBOX`, `FACEBOOK`, `GOOGLE DRIVE`, `INSTAGRAM`, `ONEDRIVE` and `ZOOM`. I don't have any option to insert an API Key for `External URL`, which is the option I might use to import files from the external OwnCloud instance.

How could I proceed? Or in this case, Uppy Companion won't be able to help me importing data from OwnCloud?

**Comment by micbar** [comment]
We added a feature to uppy companion where you can import from an ownCloud 10 public link. It is not yet upstream.

Isn't it @dschmidt ?

**Comment by dschmidt** [comment]
> We added a feature to uppy companion where you can import from an ownCloud 10 public link. It is not yet upstream.
> 
> Isn't it @dschmidt ?

Correct, upstream is working on it. I think they haven't decided whether they want to integrate it fully into their product, but they are working on providing the necessary apis at least. 

https://github.com/transloadit/uppy/pull/4621

**Comment by micbar** [comment]
Note to myself: Maybe change the ocis config to make it configurable via ENV

File: deployments/examples/ocis_wopi/.env (hunk 1 above)
[the remark this answers is not in the document]
Comment (micbar) [reply deployments/examples/ocis_wopi/.env]: i enabled only "Public Link".

**Comment by micbar** [comment]
@wkloucek @rhafer @dschmidt The three different web config files are not so nice.

Will be improved by https://github.com/owncloud/ocis/pull/8339

**Review by dragonchaser — APPROVED** [review]
lgtm

**Comment by dschmidt** [comment]
🎉 

Thanks for taking this over the finish line

**Pull request merged**